mlflow/gateway/providers/gemini.py

```diff
@@ -0,0 +1,137 @@
+                ]
+            }
+        )
+
+    @classmethod
+    def model_to_embeddings(cls, resp, config):
+        # Documentation (https://ai.google.dev/api/embeddings#v1beta.ContentEmbedding):
+        # ```
+        # {
+        #   "embeddings": [
+        #     {
+        #       "values": [
+        #         3.25,
+        #         0.7685547,
+        #         2.65625,
+        #         ...
+        #         -0.30126953,
+        #         -2.3554688,
+        #         1.2597656
+        #       ]
+        #     }
+        #   ]
+        # }
+        # ```
+
+        data = [
+            embeddings.EmbeddingObject(embedding=item.get("values", []), index=i)
+            for i, item in enumerate(resp.get("embeddings") or [resp.get("embedding", {})])
+        ]
+
+        # Create and return response payload directly
+        return embeddings.ResponsePayload(
+            data=data,
+            model=config.model.name,
+            usage=embeddings.EmbeddingsUsage(
+                prompt_tokens=None,
+                total_tokens=None,
+            ),
+        )
+
+
+class GeminiProvider(BaseProvider):
+    NAME = "Gemini"
+    CONFIG_TYPE = GeminiConfig
+
+    def __init__(self, config: RouteConfig) -> None:
+        super().__init__(config)
+        if config.model.config is None or not isinstance(config.model.config, GeminiConfig):
+            raise TypeError(f"Unexpected config type {config.model.config}")
+        self.gemini_config: GeminiConfig = config.model.config
+
+    @property
+    def headers(self):
+        return {"x-goog-api-key": self.gemini_config.gemini_api_key}
+
+    @property
+    def base_url(self):
+        return "https://generativelanguage.googleapis.com/v1beta/models"
+
+    @property
+    def adapter_class(self):
+        return GeminiAdapter
+
+    def get_endpoint_url(self, route_type: str) -> str:
+        if route_type == "llm/v1/chat" or route_type == "llm/v1/completions":
+            return f"{self.base_url}/{self.config.model.name}:generateContent"
+        elif route_type == "llm/v1/embeddings":
+            return f"{self.base_url}/{self.config.model.name}:embedContent"
+        else:
+            raise ValueError(f"Invalid route type {route_type}")
+
+    async def _request(self, path: str, payload: dict[str, Any]) -> dict[str, Any]:
+        return await send_request(
+            headers=self.headers,
+            base_url=self.base_url,
+            path=path,
+            payload=payload,
+        )
+
+    async def embeddings(self, payload: embeddings.RequestPayload) -> embeddings.ResponsePayload:
+        from fastapi.encoders import jsonable_encoder
+
+        payload = jsonable_encoder(payload, exclude_none=True)
+        self.check_for_model_field(payload)
+
+        embedding_payload = self.adapter_class.embeddings_to_model(payload, self.config)
+
+        # Use the batch endpoint if payload contains "requests"
+        if "requests" in embedding_payload:
```

Comment:
Can you add references to the relevant docs?
e.g. https://ai.google.dev/api/embeddings#EmbedContentRequest